src/commands.ts

```diff
@@ -0,0 +1,222 @@
+}
+
+type Parameters = Record<string, Variable>
+/*
+ * This command tokenizer takes the following token descriptions
+ * constant: constantValue - constant values will be verified to be the exact
+ *           value specified. If it does not match, this function will throw
+ *           InvalidConstantError
+ * variable: {type: variableName} - variable values will be verified to be the
+ *           type specified. If the type does not match, this function will throw
+ *           InvalidTypeValueError
+ * choice:   {option1|option2|option3} - choices must be one of the prespecified
+ *           values. The option will be stored in the key and value of the
+ *           returned token object. If it does not match one of the specified
+ *           options, this function will throw InvalidChoiceError
+ *
+ * types:    int, text - type must match one of these. If it does not match,
+ *           this function will throw InvalidTypeError
+ * Commands that are too short will throw TooFewTokensError
+ */
+export function tokenize(command: string, descriptions: string[]): Parameters {
+    const tokens = getTokens(_.toLower(command), descriptions.length)
+
+    const parameters: Parameters = {}
+    _.zipWith(tokens, descriptions, (token, description) => {
+        if (_.isNil(token)) {
+            throw new TooFewTokensError(tokens, descriptions)
+        } else if (description.match(VARIABLE)) {
+            parameterizeVariableToken(token, description, parameters)
+        } else if (description.match(CHOICE_VARIABLE)) {
+            parameterizeChoiceToken(token, description, parameters)
+        } else if (description.match(CONSTANT_VALUE)) {
+            parameterizeConstantToken(token, description, parameters)
+        } else {
+            throw new InvalidTokenDescriptionError(description)
+        }
+    })
+    return parameters
+}
+
+/*
+ * separates the text into numTokens-1, where the remaining words make up the final token
+ * if the number of tokens is greater than the number of " " separated words, the array of
+ * " " separated words is returned without concatenation
+ */
+function getTokens(text: string, numTokens: number) {
+    const words = _(text).split(' ').compact().value()
+    if (words.length <= numTokens) return words
+
+    const tokens = _.slice(words, 0, numTokens - 1)
+    const lastToken = _.slice(words, numTokens - 1).join(' ')
+    tokens.push(lastToken)
+
+    return tokens
+}
+
+/*
+ * get the type of the variable from the description
+ * call the specific token verification routine for the type specified
+ * get the key from the description for parameterization
+ * add the variable to the parameter object using the key from the description
+ * if the name or type cannot be determined from the dscription, throw InvalidTokenDescription
+ * if the type specified is not known, throw InvalidVariableTypeError
+ */
+function parameterizeVariableToken(
+    token: string,
+    dscription: string,
```

Comment:
Variable name - maybe call this "description" and deconstruct "type" and "name" directly?

Reply:
Fixed., Good suggestion.